lib/credentials/AssumeRoleCredentials.js

```diff
@@ -0,0 +1,68 @@
+const joi = require('@hapi/joi');
+const AWS = require('aws-sdk');
+const { Logger } = require('werelogs');
+
+const configJoi = {
+    stsConfig: joi.object().required(),
+    extension: joi.string().required(),
+    roleArn: joi.string().required(),
+    refreshCredsAnticipationSeconds: joi.number().greater(0).default(60),
+};
+
+/**
+* This class extends AWS' ChainableTemporaryCredentials class and implements
+* the needsRefresh method to refresh credentials once they are expired.
+*/
+class AssumeRoleCredentials extends AWS.ChainableTemporaryCredentials {
+    /**
+     * constructor
+     * @param {object} params - AssumeRoleCredentials params
+     * @param {string} params.extension - name of the extension
+     * @param {string} params.roleArn - ARN of the role
+     * @param {AWS.STS.ClientConfiguration} params.stsConfig - sts config
+     * @param {RequestLogger} log - request logger instance
+     * @param {number} refreshCredsAnticipationSeconds - credentials
+     * must be refreshed earlier than their effective expiration time
+     * to avoid race conditions, this sets from how many seconds
+     * before expiration time credentials are refreshed
+     */
+    constructor(params, log) {
+        const p = joi.attempt(params, configJoi);
+
+        const credentialsParams = {
+            params: {
+                RoleArn: p.roleArn,
+                RoleSessionName: `backbeat-${p.extension}`,
+                // default expiration: 1 hour,
+            },
+            stsConfig: p.stsConfig,
+        };
+
+        super(credentialsParams);
+
+        this._log = new Logger('Backbeat').newRequestLogger(log.getUids());
+        this._extension = p.extension;
+        this._refreshCredsAnticipationSeconds = p.refreshCredsAnticipationSeconds;
+    }
+
+    /**
+     * check if credentials have expired
+     * @return {boolean} - true if expired, false otherwise
+     */
+    needsRefresh() {
```

Comment:
Is overriding `needsRefresh` really necessary? It seems to be that it would be interfering with the builtin refresh mechanism with the only added value of taking into account the refresh anticipation window, which is taken care of automatically when using `.get(callback)` (even when the creds are already expired when get() is called, they are refreshed before being handed over to the callback).

Reply:
the original issue that required the anticipation window (S3C-3833) sounds like a possible issue that might occur here as well

Reply:
I may be wrong, but I think that the code in question does not use `ChainableTemporaryCredentials` and uses the credentials ad-hoc instead of obtaining then throught `.get(cb)` which indeed would result in intermittent failures.

Reply:
**update**: looking at the `aws-sdk`'s `needsRefresh`, there is already a built in expiryWindow logic.

Reply:
looking into `aws-sdk`, there is an existing mechanism in place to expire credentials 15seconds earlier, but it cannot be adjusted.
The logic here is also a copy of the aws-sdk's `needsRefresh` but with the adjustable expiry window

Reply:
My point was that the adjustable expiry window did not bring any benefit on top of builtin aws-sdk behavior (which "always" works no matter the window), it just adds another layer of timekeeping.

Reply:
> I may be wrong, but I think that the code in question does not use ChainableTemporaryCredentials and uses the credentials ad-hoc instead of obtaining them through .get(cb) before each operation, which indeed would result in intermittent failures.

`RoleCredentials` are retrieving credentials using `.get(cb)`

> My point was that the adjustable expiry window did not bring any benefit on top of builtin aws-sdk behavior (which "always" works no matter the window), it just adds another layer of timekeeping.

I have yet to see the issue in my local tests (with anticipation 0), so I understand that this fix here may be premature.

Reply:
> My point was that the adjustable expiry window did not bring any benefit on top of builtin aws-sdk behavior (which "always" works no matter the window), it just adds another layer of timekeeping.

I can remove the `needsRefresh` overwrite for now. I have yet to see an issue, so it may not be an issue.